replica/tablets.cc

```diff
@@ -440,6 +441,21 @@ tablet_id process_one_row(table_id table, tablet_map& map, tablet_id tid, const
     db_clock::time_point repair_time;
     if (row.has("repair_time")) {
         repair_time = row.get_as<db_clock::time_point>("repair_time");
+        if (db) {
+            auto myid = db->get_token_metadata().get_my_id();
+            auto range = map.get_token_range(tid);
+            auto& info = map.get_tablet_info(tid);
```

Comment:
In storage_service::replicate_to_all_cores, we do not have the info which tablet has the repair_time updated. We could update all tablets of each and every table in the system, but this is not efficient.

Reply:
We could move the update later after the map.set_tablet.


```
--- a/replica/tablets.cc
+++ b/replica/tablets.cc
@@ -439,24 +439,10 @@ future<tablet_id> process_one_row(replica::database* db, table_id table, tablet_
     }
 
     db_clock::time_point repair_time;
+    bool update_repair_time = false;
     if (row.has("repair_time")) {
         repair_time = row.get_as<db_clock::time_point>("repair_time");
-        if (db) {
-            auto myid = db->get_token_metadata().get_my_id();
-            auto range = map.get_token_range(tid);
-            auto& info = map.get_tablet_info(tid);
-            for (auto r : tablet_replicas.empty() ? info.replicas : tablet_replicas) {
-                if (r.host == myid) {
-                    co_await db->container().invoke_on(r.shard, [table, tid, range, repair_time] (replica::database& localdb) {
-                        auto& gc_state = localdb.get_compaction_manager().get_tombstone_gc_state();
-                        gc_state.update_repair_time(table, range, to_gc_clock(repair_time));
-                        tablet_logger.debug("Update repair time for tombstone gc: table={} tablet={} range={} repair_time={}",
-                                table, tid, range, repair_time);
-                    });
-                    break;
-                }
-            }
-        }
+        update_repair_time = true;
     }
 
     locator::tablet_task_info repair_task_info;
@@ -487,6 +473,23 @@ future<tablet_id> process_one_row(replica::database* db, table_id table, tablet_
 
     map.set_tablet(tid, tablet_info{std::move(tablet_replicas), repair_time, repair_task_info});
 
+    if (update_repair_time && db) {
+        auto myid = db->get_token_metadata().get_my_id();
+        auto range = map.get_token_range(tid);
+        auto& info = map.get_tablet_info(tid);
+        for (auto r : info.replicas) {
+            if (r.host == myid) {
+                co_await db->container().invoke_on(r.shard, [table, tid, range, repair_time] (replica::database& localdb) {
+                    auto& gc_state = localdb.get_compaction_manager().get_tombstone_gc_state();
+                    gc_state.update_repair_time(table, range, to_gc_clock(repair_time));
+                    tablet_logger.debug("Update repair time for tombstone gc: table={} tablet={} range={} repair_time={}",
+                            table, tid, range, repair_time);
+                });
+                break;
+            }
+        }
+    }
+
```

Reply:
> In storage_service::replicate_to_all_cores, we do not have the info which tablet has the repair_time updated. We could update all tablets of each and every table in the system, but this is not efficient.

You can propagate `state_change_hint` to replicate_to_all_cores() to optimize incremental updates.

Reply:
It's more efficient to process whole tablet_map based on state_change_hint. In your patch we ping-pong to remote shard for each updated tablet, which will be slow on full reload. We should batch it per-shard in replica_to_all_cores(), something like this:

```
db.invoke_on_all([] {
     tablet_replica this_replica = {my_id, this_shard_id()};
     return process_affected_tablets(tmap, hint, [] table_id table, tablet_info& tinfo) {
           if (has_replica(tinfo, this_replica)) {
                gc_state.update_repair_time(tinfo.repair_time);
           }
     });
});
```

Btw, if we want to keep gc_state only for tablets owned by a replica, we should also remove the entry when we cleanup tablet storage.

Reply:
> See the line below:
> 
> ```
>             auto& info = map.get_tablet_info(tid);
>             for (auto r : tablet_replicas.empty() ? info.replicas : tablet_replicas) 
> ```
> 
> it uses the latest version of tablet_replicas if it is not in info.replica yet.

Yes, but `map.get_token_range(tid)` is undefined behavior because tid may not be set in tmap (on full reload). You could hack also hack it around, but this is not clean. We should separate deserialization and updating the database structures, which happens in replica_to_all_cores(). We don't update compaction groups here too.

Reply:
> > See the line below:
> > ```
> >             auto& info = map.get_tablet_info(tid);
> >             for (auto r : tablet_replicas.empty() ? info.replicas : tablet_replicas) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it uses the latest version of tablet_replicas if it is not in info.replica yet.
> 
> Yes, but `map.get_token_range(tid)` is undefined behavior because tid may not be set in tmap (on full reload). You could also hack it around, but this is not clean. We should separate deserialization and updating the database structures, which happens in replica_to_all_cores(). We don't update compaction groups here too.

I can see that this was addressed by moving map access after set_tablet().

The other points about separation of concerns and efficiency still hold though.

Reply:
> > In storage_service::replicate_to_all_cores, we do not have the info which tablet has the repair_time updated. We could update all tablets of each and every table in the system, but this is not efficient.
> 
> You can propagate `state_change_hint` to replicate_to_all_cores() to optimize incremental updates.

Is state_change_hint a single true/false flag for all tablets? This isn't efficient either. We only need to update for the specific tablet that has repair_time changed. Calling the update storage_service::replicate_to_all_cores() is not a good place because it is not where we actually process repair_time and set it into runtime from the rows.  Isn't it better to process this in a single place, so that we do two things 1) get the repair time from rows  and set it into tablet map 2) update the repair time for tombstone gc in a central place. replicate_to_all_cores is designed to replicate token_metadata from shard zero to all shards.

Reply:
> > > See the line below:
> > > ```
> > >             auto& info = map.get_tablet_info(tid);
> > >             for (auto r : tablet_replicas.empty() ? info.replicas : tablet_replicas) 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > it uses the latest version of tablet_replicas if it is not in info.replica yet.
> > 
> > 
> > Yes, but `map.get_token_range(tid)` is undefined behavior because tid may not be set in tmap (on full reload). You could also hack it around, but this is not clean. We should separate deserialization and updating the database structures, which happens in replica_to_all_cores(). We don't update compaction groups here too.
> 
> I can see that this was addressed by moving map access after set_tablet().

Yes, after that both the range and the repair_time which are needed are ready to be used. 

> The other points about separation of concerns and efficiency still hold though.

The process_one_row is not just deserialization already.  It is the place to "translate" the tablet info from rows to runtime,  currently tablet map is the only place. Tombstone gc for repair time is just another runtime place where process_one_row could set.  I think we could even add a callback infrastructure so that we can register a callback for a specific column change in the system.tablets. 

The efficiency is a weak point. Assume we have 1M tablets in a system, 1M submit_to() calls are nothing. Also full reload should be rare. Calling tombstone gc update for only for tablet that has repair_time in process_one_row avoids the unnecessary update for unrelated tablets on a common path.

Reply:
> 1M submit_to() calls are nothing

When CPU is under load, latency can be as high as 2x task quota, so 1ms, 1M*1ms = 1000 s

Reply:
> > 1M submit_to() calls are nothing
> 
> When CPU is under load, latency can be as high as 2x task quota, so 1ms, 1M*1ms = 1000 s

If each and every submit_to takes 1ms, we will have a much bigger problem for the node to worry about. A few such calls might takes 1ms, not all of them.  In the extreme case, even if we do not call submit_to, we can also argue that each such update takes 1ms due to high load, so 1000s in total without submit to.

Reply:
> > > 1M submit_to() calls are nothing
> > 
> > 
> > When CPU is under load, latency can be as high as 2x task quota, so 1ms, 1M*1ms = 1000 s
> 
> If each and every submit_to takes 1ms, we will have a much bigger problem for the node to worry about.
> A few such calls might takes 1ms, not all of them. In the extreme case, even if we do not call submit_to, we can also argue that each such update takes 1ms due to high load, so 1000s in total without submit to.

That's not how it works. During high load, if you're processing locally, your task gets the whole task quota of 0.5ms to process tablets, and you can process many tablets in that quota, the throughput deteriorates proportionally to the load (2x load means only 2x higher processing time). With submit_to() per tablet, we're bound not by throughput but by latency, which is 2x 0.5ms (1ms) per tablet, and that's a significant slow down, updating a single tablet would take much less than that even under high load. Other things in the system would also experience latency, but throughput doesn't suffer as much because of batching, e.g. each user request submits concurrently, and we move multiple submit_to()s in a single task quota.

Reply:
Yes, indeed, submit_to is not great in this case. I added batch logic in the new version.